components/Cards/Card.tsx

```diff
@@ -1,20 +1,135 @@
-import { FC, useState, useRef, useEffect } from 'react'
+import React, { FC, useState, useRef, useEffect } from 'react'
 import {BsYoutube , BsPen} from 'react-icons/bs'
 import {AiOutlineRead} from 'react-icons/ai'
 import{MdArticle} from 'react-icons/md'
 import { CopyToClipboard } from 'components/CopyToClipboard/CopyToClipboard'
 import Share from 'components/Share/Share'
 import type { IData } from 'types'
+import { collection, doc,where,query,getDocs, setDoc,getDoc, updateDoc } from 'firebase/firestore'
+import { onAuthStateChanged } from 'firebase/auth'
+import {db,auth} from '../../lib/firebase-config'
+import { Timestamp } from 'firebase/firestore'
+import Image from 'next/image'
+import { toast } from 'react-toastify'
+import 'react-toastify/dist/ReactToastify.css';
 
 interface CardProps {
-  data: IData
+  data: IData,
 }
 
 export const Card: FC<CardProps> = ({ data }) => {
   const { name, description, url,subcategory } = data
   const descriptionRef = useRef<HTMLParagraphElement>(null)
   const [isOverflow, setIsOverflow] = useState(false)
   const youtubeRegex = /^(https?:\/\/)?(www\.)?(youtube\.com|youtu\.be)\/.+$/
+  const id = data.url.replace(/[^\w\s]/gi, '');
+  
+  const [upvoteCount,setUpvoteCount] = useState(0)
+  const [isUpvoted,setIsUpvoted] = useState(false);
+  const timestamp = Timestamp.fromDate(new Date())
+  const date = timestamp.toDate()
+  const [user,setUser] = useState<string | null>(null);
+  useEffect(() => {
+    const unsubscribe = onAuthStateChanged(auth, (user) => {
+      if (user) {
+        setUser(user.displayName);
+      } else {
+        setUser(null);
+      }
+      console.log('Authentication state changed:', user);
+    });
+  
+    return () => unsubscribe();
+  }, []);
+  const docRef = doc(db, 'resources', id)
+  const save = async()=>{
+    await setDoc(docRef, { 
+      name: name,
+      description: description,
+      url: url,
+      upvotedBy: [user],
+      upvotes: upvoteCount,
+      created: date,
+    }, 
+      { merge: true }
+    )
+  }
+
+  
+  const addUserToAssetBookmark = async () => {
+    try {
+      const subcollectionRef = collection(db, 'resources')
+      const assetQuery = query(subcollectionRef, where('name', '==', data.name))
+      const assetQuerySnapshot = await getDocs(assetQuery)
+      
+      if (assetQuerySnapshot.empty) {
+        console.log('Asset not found');
+        return;
+      }
+  
+      const assetDocSnapshot = assetQuerySnapshot.docs[0];
+      const assetDocRef = doc(db, 'resources', data.name);
+      const assetData = assetDocSnapshot.data();
+      const upvotes = assetData.upvotes || {};
+      const userUid = auth.currentUser? auth.currentUser.uid : null;
+      console.log("User ID: ",userUid)
+  
+      if (upvotes[userUid]) {
```

Comment:
@Anmol-Baranwal, according to my terminal, "Type null can't be used as an index type"(this message appears for lines 79 and 82 as well). Could this also play role in the deployment failure?